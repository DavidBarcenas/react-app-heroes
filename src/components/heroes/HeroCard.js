import React from 'react'
import { Link } from 'react-router-dom'

export const HeroCard = ({hero}) => {
  return (
    <div>
      <img src={`./assets/heroes/${hero.id}.jpg`} alt={hero.superhero} />
      <h3>{hero.superhero}</h3>
      <Link to={`/hero/${hero.id}`}>MÃ¡s...</Link>
    </div>
  )
}
