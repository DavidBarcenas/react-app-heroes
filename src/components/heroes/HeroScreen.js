import React from 'react'
import { useParams, Redirect } from 'react-router-dom'
import { getHeroesById } from '../../selectors/getHeroById';

export const HeroScreen = ({history}) => {

  const { id } = useParams();
  const hero = getHeroesById(id);

  if(!hero) { return <Redirect to='/' /> }


  const handleReturn = () => {
    if(history.length <= 2) {
      history.push('/')
    } else {
      history.goBack();
    }
  }
  return (
    <div>
      <h1>{hero.superhero}</h1>
      <button onClick={handleReturn}>Regresar</button>
    </div>
  )
}
