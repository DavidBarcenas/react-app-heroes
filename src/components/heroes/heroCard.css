.herocard {
  border: .5em solid #fff;
  max-height: 400px;
  overflow: hidden;
}

.herocard img {
  width: 100%;
  max-width: 100%;
  object-fit: cover;
}