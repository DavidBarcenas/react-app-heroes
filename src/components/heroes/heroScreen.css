.hero__wrapper {
  display: flex;
  justify-content: center;
}

.hero__img {
  width: 30%;
}

.hero__img img {
  width: 100%;
}

.hero__data {
  width: 65%;
  margin-left: 2em;
  color: #fff;
}

.hero__title {
  font-weight: bold;
  text-transform: uppercase;
  letter-spacing: .2em;
  font-size: 150px;
  word-wrap: break-word;
}