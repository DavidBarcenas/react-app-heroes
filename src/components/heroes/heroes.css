.heroes__wrapper {
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-gap: 15px;
  overflow: hidden;
  width: 100%;
}

@media screen and (min-width: 980px) {
  .heroes__wrapper {
    grid-template-columns: repeat(5, 1fr);
  }
}