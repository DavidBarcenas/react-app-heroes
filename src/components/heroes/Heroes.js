import React, { useMemo } from 'react'
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher';
import { HeroCard } from './HeroCard';

export const Heroes = ({ publisher }) => {

  const heroes = useMemo(() => getHeroesByPublisher(publisher), [publisher])

  return (
    <div>
      {
        heroes.map(hero => (
          <HeroCard hero={hero} key={ hero.id } />
        ))
      }
    </div>
  )
}
