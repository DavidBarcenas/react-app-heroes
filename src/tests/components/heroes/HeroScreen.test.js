import '@testing-library/jest-dom';
import React from 'react';
import { mount } from 'enzyme';
import { HeroScreen } from '../../../components/heroes/HeroScreen';
import { MemoryRouter, Route } from 'react-router-dom';

describe('<HeroScreen /> testing', () => {
  global.scrollTo = jest.fn()

  const historyMock = {
    length: 10,
    push: jest.fn(),
    goBack: jest.fn(),
  }

  test('should show redirect component if there are no arguments', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero']}>
        <HeroScreen history={historyMock} />
      </MemoryRouter>
    )
      expect(wrapper).toMatchSnapshot()
      expect(wrapper.find('Redirect').exists()).toBeTruthy()
  })

  test('should show a hero if the parameter exists', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-spider']}>
        <Route path="/hero/:id" component={HeroScreen} />
      </MemoryRouter>
    )
    expect(wrapper.find('.hero__wrapper').exists()).toBeTruthy()
  })
  
  test('it should return to the previous screen [Push]', () => {
    const historyMock = {
      length: 1,
      push: jest.fn(),
      goBack: jest.fn(),
    }

    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-spider']}>
        <Route path="/hero/:id" component={ () => <HeroScreen history={historyMock} /> } />
      </MemoryRouter>
    )
    wrapper.find('button').prop('onClick')();
    expect(historyMock.push).toHaveBeenCalledWith('/')
    expect(historyMock.goBack).not.toHaveBeenCalled()
  })

  test('it should return to the previous screen [goBack] ', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-spider']}>
        <Route path="/hero/:id" component={ () => <HeroScreen history={historyMock} /> } />
      </MemoryRouter>
    )
    wrapper.find('button').prop('onClick')();
    expect(historyMock.goBack).toHaveBeenCalled()
    expect(historyMock.push).not.toHaveBeenCalled()
  })
  
})
