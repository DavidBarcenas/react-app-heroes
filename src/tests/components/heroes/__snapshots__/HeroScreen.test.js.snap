// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeroScreen /> testing should show redirect component if there are no arguments 1`] = `""`;
